apps/microtvm/zephyr/template_project/src/mlperftiny/submitter_implemented.cc

```diff
@@ -0,0 +1,208 @@
+    else
+      temp_int[i] = (int8_t)(temp_uint[i] - 128);
+  }
+  g_input_data = reinterpret_cast<void*>(temp_int);
+#else
+#error Wrong model
+#endif
+}
+
+#if TARGET_MODEL == 3  // model AD
+// calculate |output - input| for AD model
+static float calculate_result() {
+  size_t feature_size = g_output_data_len;
+  float diffsum = 0;
+  float* input_float = reinterpret_cast<float*>(g_input_data);
+  float* output_float = reinterpret_cast<float*>(g_output_data);
+
+  for (size_t i = 0; i < feature_size; i++) {
+    float diff = output_float[i] - input_float[i];
+    diffsum += diff * diff;
+  }
+  diffsum /= feature_size;
+
+  return diffsum;
+}
+#endif
+
+// Add to this method to return real inference results.
+void th_results() {
+  /**
+   * The results need to be printed back in exactly this format; if easier
+   * to just modify this loop than copy to results[] above, do that.
+   */
+#if TARGET_MODEL == 3  // model AD
+  th_printf("m-results-[%0.3f]\r\n", calculate_result());
+#else
+  size_t kCategoryCount = g_output_data_len;
+  th_printf("m-results-[");
+  for (size_t i = 0; i < kCategoryCount; i++) {
+    float converted = static_cast<float>(g_quant_scale * (g_output_data[i] - g_quant_zero));
+    // float converted = static_cast<float>(g_output_data[i]);
+    th_printf("%.3f", converted);
+    if (i < (kCategoryCount - 1)) {
+      th_printf(",");
+    }
+  }
+  th_printf("]\r\n");
+#endif
+}
+
+// Implement this method with the logic to perform one inference cycle.
+void th_infer() { TVMInfer(g_input_data); }
+
+/// \brief optional API.
+void th_final_initialize(void) { TVMRuntimeInit(); }
+
+void th_pre() {}
+void th_post() {}
+
+void th_command_ready(char volatile* p_command) {
+  p_command = p_command;
+  ee_serial_command_parser_callback((char*)p_command);
+}
+
+// th_libc implementations.
+int th_strncmp(const char* str1, const char* str2, size_t n) { return strncmp(str1, str2, n); }
+
+char* th_strncpy(char* dest, const char* src, size_t n) { return strncpy(dest, src, n); }
+
+size_t th_strnlen(const char* str, size_t maxlen) { return strlen(str); }
+
+char* th_strcat(char* dest, const char* src) { return strcat(dest, src); }
+
+char* th_strtok(char* str1, const char* sep) { return strtok(str1, sep); }
+
+int th_atoi(const char* str) { return atoi(str); }
+
+void* th_memset(void* b, int c, size_t len) { return memset(b, c, len); }
+
+void* th_memcpy(void* dst, const void* src, size_t n) { return memcpy(dst, src, n); }
+
+/* N.B.: Many embedded *printf SDKs do not support all format specifiers. */
+int th_vprintf(const char* format, va_list ap) { return vprintf(format, ap); }
+void th_printf(const char* p_fmt, ...) {
+  char buffer[128];
+  int size;
+  va_list args;
+  va_start(args, p_fmt);
+  size = TVMPlatformFormatMessage(buffer, 128, p_fmt, args);
+  va_end(args);
+  TVMPlatformWriteSerial(buffer, (size_t)size);
+}
+
+char th_getchar() { return TVMPlatformUartRxRead(); }
+
+void th_serialport_initialize(void) {
```

Comment:
Would love to have comments clarifying which things we modified here and which functions are untouched from MLPerf.

Reply:
done